cmd/tiller-proxy/internal/handler/apprepos_handler.go

```diff
@@ -17,18 +17,148 @@ limitations under the License.
 package handler
 
 import (
+	"encoding/json"
+	"io"
 	"net/http"
 
 	log "github.com/sirupsen/logrus"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+
+	"github.com/kubeapps/kubeapps/cmd/apprepository-controller/pkg/apis/apprepository/v1alpha1"
+	clientset "github.com/kubeapps/kubeapps/cmd/apprepository-controller/pkg/client/clientset/versioned"
+	"github.com/kubeapps/kubeapps/pkg/auth"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// AppRepositories handles http requests for operating on app repositories
+// appRepositories handles http requests for operating on app repositories
 // in Kubeapps, without exposing implementation details to 3rd party integrations.
-type AppRepositories struct{}
+type appRepositoriesHandler struct {
+	// The config set internally here cannot be used on its own as a valid
+	// token is required. Call-sites use ConfigForToken to obtain a valid
+	// config with a specific token.
+	config rest.Config
+
+	// The namespace in which (currently) app repositories are created.
+	kubeappsNamespace string
+
+	// clientsetForConfig is a field on the struct only so it can be switched
+	// for a fake version when testing. NewAppRepositoryHandler sets it to the
+	// proper function below so that production code always has the real
+	// version (and since this is a private struct, external code cannot change
+	// the function).
+	clientsetForConfig func(*rest.Config) (clientset.Interface, error)
+}
+
+// appRepositoryRequest is used to parse the JSON request
+type appRepositoryRequest struct {
+	AppRepository appRepositoryRequestDetails `json:"appRepository"`
+}
+
+type appRepositoryRequestDetails struct {
+	Name string `json:"name"`
+	URL  string `json:"url"`
+	// TODO(mnelson): Add credential support for private repositories
+	// so this can be used for the UI request also.
+}
+
+// NewAppRepositoriesHandler returns an AppRepositories handler configured with
+// the in-cluster config but overriding the token with an empty string, so that
+// ConfigForToken must be called to obtain a valid config.
+func NewAppRepositoriesHandler(kubeappsNamespace string) (*appRepositoriesHandler, error) {
+	clientConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		clientcmd.NewDefaultClientConfigLoadingRules(),
+		&clientcmd.ConfigOverrides{
+			AuthInfo: clientcmdapi.AuthInfo{
+				Token:     "",
+				TokenFile: "",
+			},
+		},
+	)
+	config, err := clientConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+	return &appRepositoriesHandler{
+		config:             *config,
+		kubeappsNamespace:  kubeappsNamespace,
+		clientsetForConfig: clientsetForConfig,
```

Comment:
I am not sure why this function is within the `appRepositoriesHandler` struct? As far as I can see it can be called as a normal method (it doesn't depend on the struct properties).

[edit] Got it, you need this because of the tests, to be able to override it. (A comment here explaining it would be nice?)

Reply:
I added a `// See comment in the struct defn above` as there's a 5-liner explanation there :)